packages/react/src/components/Slider/Slider.js

```diff
@@ -176,41 +209,56 @@ export default class Slider extends PureComponent {
       evt.persist();
     }
 
-    if (this.state.dragging) {
-      return;
-    }
-    this.setState({ dragging: true });
-
-    this.handleDrag();
-
-    requestAnimationFrame(() => {
-      this.setState((prevState, props) => {
-        // Note: In FF, `evt.target` of `mousemove` event can be `HTMLDocument` which doesn't have `classList`.
-        // One example is dragging out of browser viewport.
-        const fromInput =
-          evt &&
-          evt.target &&
-          evt.target.classList &&
-          evt.target.classList.contains('bx-slider-text-input');
-        const { left, newValue: newSliderValue } = this.calcValue(
-          evt,
-          prevState,
-          props
-        );
-        const newValue = fromInput ? Number(evt.target.value) : newSliderValue;
-        if (prevState.left === left && prevState.value === newValue) {
-          return { dragging: false };
+    const setStateUpdater = (prevState, props) => {
+      // Note: In FF, `evt.target` of `mousemove` event can be `HTMLDocument` which doesn't have
+      // `classList`. One example is dragging out of browser viewport.
+      const fromInput =
+        evt &&
+        evt.target &&
+        evt.target.classList &&
+        evt.target.classList.contains('bx-slider-text-input');
+
+      const { left, newValue: newSliderValue } = this.calcValue(
+        evt,
+        prevState,
+        props
+      );
+
+      const newValue = fromInput ? Number(evt.target.value) : newSliderValue;
+
+      if (prevState.left === left && prevState.value === newValue) {
+        return;
+      }
+
+      return {
+        left,
+        value: newValue,
+        needsOnChange: true,
+      };
+    };
+
+    const setStateCallback = () => {
+      this.handleDragComplete();
+
+      requestAnimationFrame(() => {
```

Comment:
Seems like the call to rAF is moved from enqueuing the state update to now calling `props.onChange`, was there any particular reason for this change? Super curious 👀

Reply:
I did this primarily because the way rAF was being used was the cause of the bug, but also because it was hard to determine exactly why rAF was being used. I ran under the assumption that it was being used to rate limit the number of `onChange` and `onRelease` callbacks that are fired, since when we chatted on Webex, I believe you indicated that this type of rAF usage wasn't a common paradigm across components.

I got a message earlier from @asudoh indicating that the reason for this rAF usage was in fact to throttle the number of mousemove events that are considered for state updates, but in spite of this intent, I don't believe this is actually the way the component was behaving. Since pretty much every mousemove event ultimately resulted in a call to rAF, these events were simply being queued for batch execution at the next animation frame; none of them would actually be discarded from consideration altogether. In theory, the net number of state updates would be the same, except in the old implementation, the `setState` calls would be slightly delayed until the next animation frame picked them up for execution. This delayed state update is what causes the bug, since a mouseup event could actually fire and complete before the next animation frame had a chance to update the state in the proper way to cause the `onRelease` callback to fire.

so I guess my question is:

What is the Carbon-wide paradigm here? Are callbacks and events typically rate limited? If so, are there typical ways in which components accomplish this?

Reply:
Thank you for your thoughts @jdharvey-ibm! Basically, we use `lodash.debounce` or `lodash.throttle` whenever handling every event does not make sense; One of the cases is stale event, e.g. if we have two events to open timer-based tooltip in a short amount of time, the earlier one is stale (given the timer should be based on the latter). Another is many same events (e.g. `mousemove` or `resize`) in short amount of time, especially if the event handler involves heavy operation, which is the case of `<Slider>` (more below).

The current `<Slider>` code cancels running the logic of moving the thumb if one for earlier event is pending, which means an optimization similar to throttling:
https://github.com/carbon-design-system/carbon/blob/v10.9.0/packages/react/src/components/Slider/Slider.js#L169-L171

The direct cause of the problem is that running `onRelease()` callback is canceled if the operation of moving the thumb is canceled, even though calling `onRelease()` is _not_ considered heavy. That said, moving `handleDrag()` to before that line at least ensures that `onRelease()` is called. Wondering it's something you want to try instead of doing some refactoring.

An indirect cause of the problem that such throttling logic is written in a way that may confuse people who try to jump in and maintain the code that the logic is for managing "dragging state" instead. That caused the one who introduced `onRelease()` to put his code into a wrong place. That said, in future, I think we need to put part of `mousemove` handler code (only the logic calculating the position) into `lodash.throttle` callback rather than maintaining our own logic of throttling.

Reply:
Unfortunately making that particular change you suggest (I did indeed try that earlier on) regresses the deterministic behavior of the order in which the events are fired. Since `onChange` is still fired within the rAF call, it becomes possible to have `onRelease` called *before* `onChange`. If that is acceptable, then we can go with that instead, but I feel like it's kind of weird to change that behavior.

If the long-term goal is to convert this to use `_.throttle`, then I could probably take a stab at that over the next week or two, but I'd like your feedback on that before going off in that direction.

Reply:
`onChange`/`onRelease` order is a great point. A quick fix may be enqueueing `onRelease` call instead of calling it directly at https://github.com/carbon-design-system/carbon/blob/v10.9.0/packages/react/src/components/Slider/Slider.js#L347.

And you are right that the long-term goal is converting the code to use `_.throttle`.